### PR TITLE
Add visits to nodes that already have flat_maps in ast::MutVisitor

### PR DESCRIPTION
**Pull request by maxcabrajac**
This PR aims to add `visit_` methods for every node that has a `flat_map_` in MutVisitor, giving implementers free choice over overriding `flat_map` for 1-to-n conversions or `visit` for a 1-to-1.

There is one major problem: `flat_map_stmt`.
While all other default implementations of `flat_map`s are 1-to-1 conversion, as they either only call visits or a internal 1-to-many conversions are natural, `flat_map_stmt` doesn't follow this pattern.

`flat_map_stmt`'s default implementation is a 1-to-n conversion that panics if n > 1 (effectively being a 1-to-[0;1]). This means that it cannot be used as is for a default `visit_stmt`, which would be required to be a 1-to-1. 

Implementing `visit_stmt` without runtime checks would require it to reach over a potential `flat_map_item` or `filter_map_expr` overrides and call for their `visit` counterparts directly.
Other than that, if we want to keep the behavior of `flat_map_stmt` it cannot call `visit_stmt` internally.

To me, it seems reasonable to make all default implementations 1-to-1 conversions and let implementers handle `visit_stmt` if they need it, but I don't know if calling `visit` directly when a 1-to-1 is required is ok or not. 

related to #128974 & #127615 

r? @petrochenkov

### CODE REVIEW

File: compiler/rustc_builtin_macros/src/test_harness.rs [hunk 1]
```diff
@@ -129,8 +130,8 @@ impl<'a> MutVisitor for TestHarnessGenerator<'a> {
         c.items.push(mk_main(&mut self.cx));
     }
 
-    fn flat_map_item(&mut self, mut i: P<ast::Item>) -> SmallVec<[P<ast::Item>; 1]> {
-        let item = &mut *i;
+    fn visit_item(&mut self, i: &mut P<ast::Item>) {
+        let item = i.deref_mut();
```
Comment (petrochenkov) [inline compiler/rustc_builtin_macros/src/test_harness.rs:134]: Could you pick up the test harness cleanups from https://github.com/rust-lang/rust/pull/127371?
The `deref_mut`s shouldn't be necessary in particular.

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 2]
```diff
@@ -379,21 +379,21 @@ fn index_crate<'a>(
         fn visit_item(&mut self, item: &'a ast::Item) {
             let def_id = self.node_id_to_def_id[&item.id];
             *self.index.ensure_contains_elem(def_id, || AstOwner::NonOwner) = AstOwner::Item(item);
-            visit::walk_item(self, item)
+            visit::walk_item(self, item, ())
```
Comment (petrochenkov) [inline compiler/rustc_ast_lowering/src/lib.rs:382]: The `()` stuff is an implementation detail that shouldn't generally leak outside of the (mut) visit modules, IMO.
There should be a wrapper hiding it, and `walk_item` was such a wrapper (probably still can be).

Comment (maxcabrajac) [reply compiler/rustc_ast_lowering/src/lib.rs]: I changed the general (with ctxt) function from `walk_item` to `walk_item_ctxt` and added `walk_item` and `walk_assoc_item` wrappers

File: compiler/rustc_builtin_macros/src/test_harness.rs (hunk 1 above)
Comment (maxcabrajac) [reply compiler/rustc_builtin_macros/src/test_harness.rs]: There is one place where an intermediate unwrapped (`&mut Item` vs `&mut P<Item>`) is still needed. There we need multiple mut refs to fields of `Item` to pass to `walk_item_kind`. I swapped it to `&mut **item` so we don't include `DerefMut`

**Comment by maxcabrajac** [comment]
Should I make a first test implementation of what I think should be done with `flat_map_stmt`? Do you have any opinions on it?

**Comment by petrochenkov** [comment]
The current changes are good to go, after cleaning up history to avoid the back and forth changes.
The statement visiting can be done separately so I can think about it later, and not right now.

**Comment by maxcabrajac** [comment]
Ok =)
History cleaned.
I'll open another pr with a first implementation of statement visiting.

**Pull request merged**